File: Calorimeter/CaloClusterCorrection/python/CaloSwPhimod_g3.py
# Copyright (C) 2002-2017 CERN for the benefit of the ATLAS collaboration

#
# $Id: CaloSwPhimod_g3.py,v 1.1 2006-11-14 14:51:28 ssnyder Exp $
#
# File: CaloClusterCorrection/python/CaloSwPhimod_g3.py
# Created: Nov 2006, sss
# Purpose: Phi modulation corrections, original G3 version.
#
# This is the original G3-based phi modulation correction, translated from the
# original fortran version.
#


class CaloSwPhimod_g3_parms:
    correction = [
        [0.1, 0.9998,  0.0049,  0.0004,  0.0004,  0.0021],
        [0.3, 0.9942,  0.0060, -0.0001, -0.0001,  0.0026],
        [0.5, 0.9923,  0.0058, -0.0010, -0.0005,  0.0041],
        [0.7, 0.9942,  0.0055, -0.0005, -0.0006,  0.0026],
        [0.9, 0.9939,  0.0044,  0.0006,  0.0000, -0.0012],
        [1.2, 0.9878,  0.0026,  0.0000,  0.0003, -0.0021],
        [1.5, 1.0,     0.0,     0.0,     0.0,     0.0],
        [1.7, 0.9992, -0.0005, -0.0032,  0.0038, -0.0002],
        [1.8, 0.9966, -0.0038,  0.0004,  0.0033,  0.0001],
        [1.9, 0.9955, -0.0082,  0.0000,  0.0022,  0.0016],
        [2.0, 0.9935, -0.0115,  0.0026,  0.0003,  0.0015],
        [2.1, 0.9934, -0.0107,  0.0018,  0.0008,  0.0017],
        [2.2, 0.9938, -0.0095,  0.0003,  0.0010,  0.0009],
        [2.3, 0.9954, -0.0043,  0.0,     0.0013,  0.0005],
        [2.5, 1.0,     0.0,     0.0,     0.0,     0.0],
        ]

File: Calorimeter/CaloClusterCorrection/src/CaloSwPhimod_g3.cxx
/*
  Copyright (C) 2002-2020 CERN for the benefit of the ATLAS collaboration
*/

/********************************************************************

NAME:     CaloSwPhimod_g3.cxx
PACKAGE:  offline/Calorimeter/CaloClusterCorrection

AUTHORS:  H. Ma, S. Rajagopalan
CREATED:  Dec. 15, 1999

PURPOSE:  correction for the modulations with the position in phi
          (Tuned using 50 GeV Et photons)  
          base class: CaloClusterCorrection
          Correction tuned on G3 samples.

Atrecon Orig: emreco/qphimod.F

Updated:  May 10, 2000    (SR, HM)
          Migrated to Athena Framework from PASO

Updated:  Jan 5, 2001    (HM)
          QA. 

Updated:  May 5, 2004    (Sven Menke)
	  base class changed from algo to tool

Updated:  June, 2004    (sss)
          Use ToolWithConstants to get correction constants.
********************************************************************/
#include "CaloSwPhimod_g3.h"
#include <cmath>



// make correction to one cluster 
void CaloSwPhimod_g3::makeCorrection (const Context& myctx,
                                      xAOD::CaloCluster* cluster) const
{

  float eta2 = cluster->etaBE(2); // use second sampling
  float phi2 = cluster->phiBE(2); // use the average over Bar and EC.
  if (eta2 < -900 || phi2 < -900)
    return;

  float aeta2 = fabs(eta2);
  float nabs;
  if (aeta2 < 1.5) 
      nabs = 1024; 
  else if (aeta2 < 2.5)
      nabs = 768;
  else
      nabs = 256.;

  std::vector<float> tab = qphmop (myctx, aeta2);
  assert (tab.size() == 5);
  float qphimod = 
    tab[0] + 
    tab[1] * cos(nabs*phi2) +
    tab[2] * cos(2.*nabs*phi2)+
    tab[3] * cos(3.*nabs*phi2)+ 
    tab[4] * sin(nabs*phi2);
  
  // set energy, and rescale each sampling
  setenergy(cluster, qphimod * cluster->e());
  
  
}

std::vector<float> CaloSwPhimod_g3::qphmop (const Context& myctx,
                                            float aeta) const
{
  const CxxUtils::Array<2> correction = m_correction (myctx);

  std::vector<float> coef (correction.size(1)-1);

  int i1 = 0;
  int i2 = 0;
  if (aeta < correction[0][0]) {
    for (unsigned int j = 1; j < correction.size(1); j++)
      coef[j-1] = correction[0][j];
    return coef;
  }
    
  else if (aeta >= correction[0][0]) {
    for (unsigned int i = 0; i < correction.size()-1; i++) {
      if (aeta >= correction[i][0] && aeta < correction[i+1][0]) {
        i1 = i;
        i2 = i+1;
        break;
      }
    }
  }
  if (i1 == i2) {
    i2 = correction.size()-1;
    i1 = i2-1;
  }

  float deta = correction[i2][0] - correction[i1][0];
  if (deta == 0)
    deta = 1;

  float m = (aeta - correction[i1][0]) / deta;
  for (unsigned int j = 1; j < correction.size(1); j++) {
    float dy = correction[i2][j] - correction[i1][j];
    coef[j-1] = (correction[i1][j] + dy * m);
  }

  return coef;
}

